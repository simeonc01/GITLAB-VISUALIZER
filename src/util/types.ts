// gitlab types
export type Author = {
    state: string;
    id: number;
    web_url: string;
    name: string;
    avatar_url: string | null;
    username: string;
}

export type Reference =  {
    short: string;
    relative: string;
    full: string;
}

export type Milestone = {
    created_at: Date;
    description: string;
    due_date: string | null;
    expired: boolean;
    id: number;
    iid: number;
    project_id: string;
    start_date: string;
    state: string;
    title: string;
    updated_at: Date;
    web_url: string;
}

export type Commit = {
    author_email: string;
    author_name: string;
    authored_date: string;
    committed_date: string;
    committer_email: string;
    committer_name: string;
    created_at: string;
    id: string;
    message: string;
    parent_ids: string[];
    short_id: string;
    title: string;
    trailers: object;
    web_url: string;
}

export type Branch = {
    can_push: boolean;
    commit: Commit;
    default: boolean;
    developers_can_merge: boolean;
    developers_can_push: boolean;
    merged: boolean;
    name: string;
    protected: boolean;
    web_url: string;
}

export type Issue = {
    _links: object;
    assignee: Author;
    assignees: Author[];
    author: Author;
    closed_at: Date | null;
    closed_by: Author;
    confidential: boolean;
    created_at: Date;
    description: string;
    discussion_locked: boolean | null;
    downvotes: number;
    due_date: string | null;
    has_tasks: boolean;
    id: number;
    iid: number;
    issue_type: string;
    labels: string[];
    merge_request_count: number;
    milestone: Milestone;
    moved_to_id: number | null;
    project_id: number;
    references: Reference;
    service_desk_reply_to: string | null;
    severity: string;
    state: string;
    task_completion_status: object;
    time_stats: object;
    title: string;
    type: string;
    updated_at: Date;
    upvotes: number;
    user_notes_count: number;
    web_url: string;
}

export type Project = {
    _links: object;
    allow_merge_on_skipped_pipeline: any;
    analytics_access_level: string;
    archived: boolean;
    auto_cancel_pending_pipelines: string;
    auto_devops_deploy_strategy: string;
    auto_devops_enabled: boolean;
    autoclose_referenced_issues: boolean;
    avatar_url: string | null;
    build_timeout: number;
    builds_access_level: string;
    can_create_merge_request_in: boolean;
    ci_allow_fork_pipelines_to_run_in_parent_project: boolean;
    ci_config_path: string;
    ci_default_git_depth: number;
    ci_forward_deployment_enabled: boolean;
    ci_job_token_scope_enabled: boolean;
    ci_opt_in_jwt: boolean;
    ci_separated_caches: boolean;
    container_expiration_policy: object;
    container_registry_access_level: string;
    container_registry_enabled: boolean;
    container_registry_image_prefix: string;
    created_at: string;
    creator_id: number;
    default_branch: string;
    description: string | null;
    emails_disabled: string | null;
    empty_repo: boolean;
    enforce_auth_checks_on_uploads: boolean;
    forked_from_project: object;
    forking_access_level: string;
    forks_count: number;
    http_url_to_repo: string;
    id: number;
    import_status: string;
    issues_access_level: string;
    issues_enabled: string;
    jobs_enabled: string;
    keep_latest_artifact: string;
    last_activity_at: string;
    lfs_enabled: boolean;
    merge_commit_template: string | null;
    merge_method: string;
    merge_requests_access_level: string;
    merge_requests_enabled: boolean;
    mr_default_target_self: boolean;
    name: string;
    name_with_namespace: string;
    namescape: object;
    only_allow_merge_if_all_discussions_are_resolved: boolean;
    only_allow_merge_if_pipeline_succeeds: boolean;
    open_issues_count: number;
    operations_access_level: string;
    owner: Author;
    packages_enabled: boolean;
    pages_access_level: string;
    path: string;
    path_with_namespace: string;
    permissions: object;
    printing_merge_request_link_enabled: boolean;
    public_jobs: boolean;
    readme_url: string | null;
    remove_source_branch_after_merge: boolean;
    repository_access_level: string;
    request_access_enabled: boolean;
    resolve_outdated_diff_discussions: boolean;
    restrict_user_defined_variables: boolean;
    runner_token_expiration_interval: string | null;
    security_and_compliance_access_level: string;
    service_desk_enabled: boolean;
    shared_runners_enabled: boolean;
    shared_with_groups: [];
    snippets_access_level: string;
    snippets_enabled: boolean;
    squash_commit_template: string | null;
    squash_option: string;
    ssh_url_to_repo: string;
    star_count: number;
    suggestion_commit_message: string | null;
    tag_list: [];
    topics: [];
    visibility: string;
    web_url: string;
    wiki_access_level: string | null;
    wiki_enabled: boolean;
}


export type Event = {
    id: number;
    title: string | null;
    project_id: number;
    action_name: string;
    target_id: number;
    target_type: string;
    author_id: number;
    target_title: string;
    created_at: string;
    author: Author;
    author_username: string;
}

export type Label = {
    id: number;
    name: string;
    color: string;
    text_color: string;
    description: string;
    description_html: string;
    open_issue_count: number;
    closed_issues_count: number;
    open_merge_requests_count: number;
    subscribed: boolean;
    priority: string | null;
    is_project_label: boolean;
}

// other types
export interface IContextDefault {
    commits: Commit[] | null;
    branches: Branch[] | null;
    issues: Issue[] | null;
    labels: Label[] | null;
    currentProject: Project | null;
    events: Event[] | null;
    error: boolean;
    loading: boolean;
    update: () => void;
}

export type UpdateData = {
    commits: Commit[];
    branches: Branch[];
    issues: Issue[];
    currentProject: Project;
    events: Event[];
    labels: Label[];
}

export type GitlabError = {
    message: string;
    data: object | null;
}

// recharts types
export type LineType = {
    dataKey: string;
    stroke?: string;
}

export type FilterType = {
    startDate: Date | null;
    endDate: Date | null;
}