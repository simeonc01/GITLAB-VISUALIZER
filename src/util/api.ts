import axios, { AxiosInstance, AxiosPromise, AxiosResponseHeaders } from 'axios';
import { Branch, Commit, Issue, Project, UpdateData, Event, Label } from './types';

function wrapPromise<T>(axios: AxiosPromise<T>) {
    return new Promise<T>((resolve, reject) => {
        axios
            .then((d) => {
                resolve(d.data);
            })
        .catch(async (error) => {
            if (error.response) {
                reject(error.response.data.message);
            } else {
                reject(error.message);
            }
        });
    });
}

function wrapPromiseHeaderIncluded<T>(axios: AxiosPromise<T>) {
    return new Promise<{data: T, headers: AxiosResponseHeaders}>((resolve, reject) => {
        axios.then(d => {
            resolve({data: d.data, headers: d.headers});
        }).catch(async (error) => {
            if (error.response) {
                reject(error.response.data.message);
            } else {
                reject(error.message);
            }
        });
    })
}




export class ApiHandler {
    private readonly handler: AxiosInstance;
    private id = -1;
    private projectString: string;
    private originalProjectString: string;
    private initialized = false;
    private token: string;

    constructor(token: string, projectString: string) {
        this.originalProjectString = projectString;
        const r = projectString.match(/(?<=\.no\/)[^\]]+/);
        this.projectString = encodeURIComponent(r !== null ? r[0] : "");
        this.token = token;
        this.handler = axios.create({
            baseURL: `https://gitlab.stud.idi.ntnu.no/api/v4`,
            timeout: 3000
        });
        this.handler.defaults.headers.common["PRIVATE-TOKEN"] = token;
    }

    public async update(): Promise<UpdateData> {
        return new Promise<UpdateData>((resolve, reject) => {
            Promise.all([
                this.getCommits(),
                this.getBranches(),
                this.getIssues(),
                this.getCurrentProject(),
                this.getEvents(),
                this.getLabels()
            ]).then(data => {
                resolve({
                    commits: data[0],
                    branches: data[1],
                    issues: data[2],
                    currentProject: data[3],
                    events: data[4],
                    labels: data[5]
                });
            }).catch(error => {
                reject({
                    message: error,
                    data: null
                });
            });
        })
    }

    private async getProjectId(projectString: string): Promise<number> {
        return new Promise<number>((resolve, reject) => {
            fetch(`https://gitlab.stud.idi.ntnu.no/api/v4/projects/${encodeURIComponent(projectString)}`, {
                headers: {
                    'PRIVATE-TOKEN': this.token
                }
            }).then(response => {
                if (response.ok) {
                    response.json().then((d: Project) => resolve(d.id));
                } else
                    reject(-1);
            })

        });
    }

    public async getCurrentProject(): Promise<Project> {
        if (this.id < 0)
            return Promise.reject<Project>({
                message: "Project ID was not set", 
                data: null
            });
        return wrapPromise<Project>(this.handler.get(`/projects/${this.id}`, {
            validateStatus: (status) => status === 200
        }));
    }

    public async updateDetails(token: string, projectString: string): Promise<boolean> {
        if (this.token === token && this.originalProjectString === projectString) return Promise.resolve<boolean>(true);
        this.token = token;
        this.handler.defaults.headers.common["PRIVATE-TOKEN"] = this.token;
        const r = projectString.match(/(?<=\.no\/)[^\]]+/);
        if (r !== null) this.projectString = r[0];
        else this.projectString = "";
        const id = await this.getProjectId(this.projectString);
        if (id < 0) Promise.resolve<boolean>(false);
        this.id = id;
        return Promise.resolve<boolean>(true);
    }

    public async getProjects(page?: number): Promise<Project[]> {
        if (page) {
            return wrapPromise<Project[]>(this.handler.get(`/projects?per_page=100&page=${page}`));
        }

        return wrapPromise<Project[]>(this.handler.get(`/projects?per_page=100`));
    }

    public async getCommits(): Promise<Commit[]> {
        if (this.id < 0)
            return Promise.reject<Commit[]>({
                message: "Project ID was not set", 
                data: null
            });
        
        try {
            const data = await wrapPromiseHeaderIncluded<Commit[]>(
                this.handler.get(`/projects/${this.id}/repository/commits?per_page=100`, {
                    validateStatus: (code: number) => code === 200
                })
            )

            let iteration = 2
            let stopIteration = data.headers["x-next-page"] === "";
    
            while (!stopIteration) {
                const d = await wrapPromise(this.handler.get(`/projects/${this.id}/repository/commits?per_page=100&page=${iteration}`))
                data.data.push(...d);

                const dataHeader = await wrapPromiseHeaderIncluded<Commit[]>(
                    this.handler.get(`/projects/${this.id}/repository/commits?per_page=100&page=${iteration}`, {
                        validateStatus: (code: number) => code === 200
                    })
                )

                if (dataHeader.headers["x-next-page"] !== "") {
                    iteration++;
                }
                else {
                    stopIteration = true;
                }
            }

            
            return Promise.resolve<Commit[]>(data.data);
            
        } catch (e) {
            return Promise.reject<Commit[]>({
                message: e, 
                data: null
            });
        } 
    }

    
    public async getIssues(): Promise<Issue[]> {
        if (this.id < 0)
            return Promise.reject<Issue[]>({
                message: "Project ID was not set", 
                data: null
            });


            try {
                const data = await wrapPromiseHeaderIncluded<Issue[]>(
                    this.handler.get<Issue[]>(`/projects/${this.id}/issues?per_page=100`, {
                        validateStatus: (code: number) => code === 200
                    })
                );
        
                if (parseInt(data.headers["x-total-pages"]) > 1) {
                    for (let i = 2; i <= parseInt(data.headers["x-total-pages"]); i++) {
                        const d = await wrapPromise(this.handler.get(`/projects/${this.id}/issues?per_page=100&page=${i}`))
                        data.data.push(...d);
                    }
                }
        
                return Promise.resolve<Issue[]>(data.data);
            } catch (e) {
                return Promise.reject<Issue[]>({
                    message: e, 
                    data: null
                });
            }
    }

    public async getBranches(): Promise<Branch[]> {
        if (this.id < 0)
            return Promise.reject<Branch[]>({
                message: "Project ID was not set", 
                data: null
            });
        
            try {
                const data = await wrapPromiseHeaderIncluded<Branch[]>(
                    this.handler.get<Branch[]>(`/projects/${this.id}/repository/branches?per_page=100`, {
                        validateStatus: (code: number) => code === 200
                    })
                );
        
                if (parseInt(data.headers["x-total-pages"]) > 1) {
                    for (let i = 2; i <= parseInt(data.headers["x-total-pages"]); i++) {
                        const d = await wrapPromise(this.handler.get(`/projects/${this.id}/repository/brances?per_page=100&page=${i}`))
                        data.data.push(...d);
                    }
                }
        
                return Promise.resolve<Branch[]>(data.data);
            } catch (e) {
                return Promise.reject<Branch[]>({
                    message: e, 
                    data: null
                });
            }
    }

    public async getEvents(): Promise<Event[]> {
        if (this.id < 0)
            return Promise.reject<Event[]>({
                message: "Project ID was not set", 
                data: null
            });

        try {
            const data = await wrapPromiseHeaderIncluded<Event[]>(
                this.handler.get<Event[]>(`/projects/${this.id}/events?per_page=100`, {
                    validateStatus: (code: number) => code === 200
                })
            );
            if (parseInt(data.headers["x-total-pages"]) > 1) {
                for (let i = 2; i <= parseInt(data.headers["x-total-pages"]); i++) {
                    const d = await wrapPromise(this.handler.get(`/projects/${this.id}/events?per_page=100&page=${i}`))
                    data.data.push(...d);
                    console.log(d.length);
                }
            }
    
            return Promise.resolve<Event[]>(data.data);
        } catch (e) {
            return Promise.reject<Event[]>({
                message: e, 
                data: null
            });
        }
    }   

    public async getLabels(): Promise<Label[]> {
        if (this.id < 0)
            return Promise.reject<Label[]>({
                message: "Project ID was not set", 
                data: null
            });
        
        return wrapPromise<Label[]>(
            this.handler.get<Label[]>(`/projects/${this.id}/labels`, {
                validateStatus: (code: number) => code === 200
            })
        )
    }
}