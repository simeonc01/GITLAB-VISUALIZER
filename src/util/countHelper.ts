import { BetterCommit, BetterEvent, BetterIssue, FreeDictionary, KeyCount } from "./types";

export function countAmountPerDate (list: BetterEvent[] | BetterIssue[] | BetterCommit[]): KeyCount[] {
    const r: FreeDictionary<number> = {};
    const uniqueDates = new Set<string>();
    list.map(d => (d.created_at_date.toISOString())).forEach(a => uniqueDates.add(a));
    uniqueDates.forEach(date => r[date] = 0);
    list.forEach(d => r[d.created_at_date.toISOString()]++);
    return Object.keys(r).map(k => ({index: new Date(k).toLocaleDateString('en-us'), count: r[k]}));
}

export function countAmoutPerType(list: BetterEvent[]): KeyCount[] {
    const r: FreeDictionary<number> = {};
    const uniqueTypes = new Set<string>();
    list.forEach(a => uniqueTypes.add(a.action_name));
    uniqueTypes.forEach(a => r[a] = 0);
    list.forEach(a => r[a.action_name]++);
    return Object.keys(r).map(k => ({index: k, count: r[k]}));
}

export function countAmoutPerAuthor(list: BetterEvent[]): KeyCount[] {
    const r: FreeDictionary<number> = {};
    const uniqueAuthors = new Set<string>();
    list.forEach(a => uniqueAuthors.add(a.author.username));
    uniqueAuthors.forEach(a => r[a] = 0);
    list.forEach(a => r[a.author.username]++);
    return Object.keys(r).map(k => ({index: k, count: r[k]}));
}