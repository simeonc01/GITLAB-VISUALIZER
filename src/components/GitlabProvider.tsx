import React, { createContext, ReactNode, useEffect, useState } from 'react';
import { ApiHandler } from '../util/api';
import { Branch, Commit, GitlabError, IContextDefault, Issue, Project } from '../util/types';


export const GitLabContext = createContext<IContextDefault>({} as IContextDefault);

const GitlabProvider = (props: {children?: ReactNode}) => {
    const [commits, setCommits] = useState<Commit[]>([]);
    const [branches, setBranches] = useState<Branch[]>([]);
    const [issues, setIssues] = useState<Issue[]>([]);
    const [currentProject, setCurrentProject] = useState<Project>({} as Project);
    const [error, setError] = useState<boolean>(false);
    const [loading, setLoading] = useState<boolean>(true);

    const apiHandler = new ApiHandler("", "");

    const getCommits = (): Commit[] | null => {
        if (error) return null;
        return commits;
    }

    const getBranches = (): Branch[] | null => {
        if (error) return null;
        return branches;
    }

    const getIssues = (): Issue[] | null => {
        if (error) return null;
        return issues;
    }

    const getCurrentProject = (): Project | null => {
        if (error) return null;
        return currentProject;
    }

    const updateData = () => {
        setLoading(true);
        apiHandler.update().then(data => {
            setCommits(data.commits);
            setBranches(data.branches);
            setIssues(data.issues);
            setCurrentProject(data.currentProject);
            setLoading(false);
        }).catch((err: GitlabError) => {
            setError(true);
            console.error("Error in update");
            console.error(err.message);
        });
    }

    useEffect(() => {
        update(localStorage.getItem("token"), localStorage.getItem("projectName"));

    }, []);

    const update = async (token: string | null , projectName: string | null) => {
        if (token === null || projectName === null){
            setError(true);
            return;
        }

        const success = await apiHandler.updateDetails(token, projectName);

        if (!error && success) {
            updateData();
            localStorage.setItem("token", token);
            localStorage.setItem("projectName", projectName);
        } else
            console.error("Context is not setup correctly, need a valid Token and projectName")
    }

    return (
        <GitLabContext.Provider value={{
            commits: getCommits(),
            branches: getBranches(), 
            issues: getIssues(), 
            currentProject: getCurrentProject(), 
            error,
            loading,
            update
        }}>
            {props.children}
        </GitLabContext.Provider>
    )
}

export default GitlabProvider;
