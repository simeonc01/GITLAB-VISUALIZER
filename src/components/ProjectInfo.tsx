import {
  Box,
  CircularProgress,
  Divider,
  Tooltip,
  Typography,
} from "@mui/material";
import React, { useContext, useEffect, useState } from "react";
import { IContextDefault, Label, Milestone, Project } from "../util/types";
import { GitLabContext } from "./GitlabProvider";
import Container from "./LayoutContainer";

const LabelComponent = (props: { label: Label }) => {
  if (props.label.description !== null && props.label.description !== "") {
    return (
      <Tooltip title={props.label.description} placement="bottom">
        <Box
          style={{
            backgroundColor: props.label.color,
            color: props.label.text_color,
          }}
          borderRadius={1}
          display={"inline-block"}
        >
          <Typography variant="body1" paddingX={1}>
            {props.label.name}
          </Typography>
        </Box>
      </Tooltip>
    );
  } else {
    return (
      <Box
        sx={{
          backgroundColor: props.label.color,
          color: props.label.text_color,
          borderRadius: 1,
          display: "inline-block",
          width: "fit-content",
          margin: 0.25,
        }}
      >
        <Typography variant="body2" paddingX={1}>
          {props.label.name}
        </Typography>
      </Box>
    );
  }
};

const InfoBox = (props: {
  project: Project;
  labels: Label[];
  milestones: Milestone[];
}) => {
  return (
    <>
      <Typography
        variant="h5"
        sx={{
          textTransform: "capitalize",
        }}
      >
        {props.project.name}
      </Typography>
      <Box
        sx={{
          display: "flex",
          alignItems: "center",
          width: 1,
          flexDirection: "row",
        }}
      >
        <Box
          sx={{
            display: "flex",
            flexGrow: 1,
            flexDirection: "column",
            width: "fit-content",
            marginRight: 0.5,
          }}
        >
          <Typography variant="body1">
            ‚≠ê Your project has {props.project.star_count} stars
          </Typography>
          <Typography variant="body1">
            üí™ The last activity was at{" "}
            {new Date(props.project.last_activity_at).toDateString()}
          </Typography>
          <Typography variant="body1">
            ü•ö Your project was created at{" "}
            {new Date(props.project.created_at).toDateString()}
          </Typography>
          <Typography variant="body1">
            üç¥ Your project has been forked {props.project.forks_count} times
          </Typography>
          <Typography variant="body1">
            üö© Your project has {props.milestones.length} milestone
            {props.milestones.length === 1 ? "" : "s"}
          </Typography>
        </Box>
        <Divider orientation="vertical" flexItem />
        <Box
          sx={{
            display: "flex",
            flexDirection: "column",
            float: "right",
            alignItems: "center",
            marginLeft: 1,
          }}
        >
          <Typography variant="body1">Your projects labels:</Typography>
          {props.labels.map((l) => (
            <LabelComponent key={l.id} label={l} />
          ))}
        </Box>
      </Box>
    </>
  );
};

const ProjectInfo = () => {
  const context = useContext<IContextDefault>(GitLabContext);

  const [project, setProject] = useState<Project>();
  const [labels, setLabels] = useState<Label[]>([]);
  const [error, setError] = useState<boolean>(false);
  const [milestones, setMilestones] = useState<Milestone[]>([]);

  useEffect(() => {
    if (!context.loading) {
      if (
        context.currentProject !== null &&
        context.labels !== null &&
        context.milestones !== null
      ) {
        setProject(context.currentProject);
        setLabels(context.labels);
        setMilestones(context.milestones);
      } else setError(true);
    }
  }, [context.loading]);

  return (
    <Container>
      {context.loading && <CircularProgress />}
      {!context.loading &&
        !error &&
        project !== undefined &&
        labels.length !== 0 && (
          <InfoBox project={project} labels={labels} milestones={milestones} />
        )}
    </Container>
  );
};

export default ProjectInfo;
