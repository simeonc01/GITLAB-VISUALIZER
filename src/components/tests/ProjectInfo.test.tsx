import renderer from "react-test-renderer";
import { Label, Project } from "../../util/types";
import { GitLabContext } from "../GitlabProvider";
import ProjectInfo from "../ProjectInfo";

it("Only component test", () => {
  const p = {} as Project;
  p.description = "Description";
  p.name = "Name";
  p.star_count = 2;
  p.last_activity_at = "2022-10-04T13:52.109+02:00";
  p.created_at = "2022-09-04T13:52.109+02:00";
  p.forks_count = 0;

  const l = {} as Label;
  l.color = "#d9534f";
  l.name = "bug";
  l.text_color = "#FFFFFF";

  const list: Label[] = [];
  list.push(l);

  const comp = renderer
    .create(
      <GitLabContext.Provider
        value={{
          commits: [],
          branches: [],
          issues: [],
          currentProject: p,
          events: [],
          labels: list,
          error: false,
          loading: false,
          update: () => true,
        }}
      >
        <ProjectInfo />
        <h2>hey</h2>
      </GitLabContext.Provider>
    )
    .toJSON();
  expect(comp).toMatchSnapshot();
});

// it("Test with data", () => {
//     const comp = renderer.create(<CommitTable rows ={testData} />).toJSON();
//     expect(comp).toMatchSnapshot();
//   });

export {};
