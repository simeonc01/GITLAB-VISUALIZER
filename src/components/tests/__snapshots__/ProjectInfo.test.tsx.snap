// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple ProjectInfoContent test 1`] = `
Array [
  <h5
    className="MuiTypography-root MuiTypography-h5 css-p4tdjl-MuiTypography-root"
  >
    Name
  </h5>,
  <div
    className="MuiBox-root css-12wop9s"
  >
    <div
      className="MuiBox-root css-1rbokul"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        ⭐ Your project has 
        2
         stars
      </p>
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        💪 The last activity was at
         
        Invalid Date
      </p>
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        🥚 Your project was created at
         
        Invalid Date
      </p>
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        🍴 Your project has been forked 
        0
         times
      </p>
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        🚩 Your project has 
        0
         milestone
        s
      </p>
    </div>
    <hr
      className="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
    />
    <div
      className="MuiBox-root css-memq11"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        Your projects labels:
      </p>
      <div
        aria-label="desc"
        aria-labelledby={null}
        className="MuiBox-root css-1uc3ujw"
        data-mui-internal-clone-element={true}
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "backgroundColor": "#d9534f",
            "color": "#FFFFFF",
          }
        }
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-1ezobfm-MuiTypography-root"
        >
          bug
        </p>
      </div>
    </div>
  </div>,
]
`;
